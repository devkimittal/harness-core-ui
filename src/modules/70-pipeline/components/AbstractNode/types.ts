import type { IconName } from '@harness/uicore'

export interface ListenerHandle {
  deregister: () => any
  id: string
  listener: BaseListener
}

export interface NodeData {
  name: string
  icon: IconName
  selectedColour: string
  unSelectedColour: string
  selectedIconColour: string
  unSelectedIconColour: string
}

export type BaseListener = (event: any) => void

export const enum PipelineGraphType {
  STAGE_GRAPH = 'STAGE_GRAPH',
  STEP_GRAPH = 'STEP_GRAPH'
}
export interface PipelineGraphState {
  id: string
  identifier: string
  type: string
  name: string
  icon: IconName
  status?: string
  data: any
  nodeType?: string
  graphType?: PipelineGraphType
  children?: PipelineGraphState[]
  parentStepGroupId?: string
  readonly?: boolean
}
export interface NodeIds {
  startNode: string
  createNode: string
  endNode: string
}

export interface SVGPathRecord {
  [key: string]: string
}

export type NodeBank = Map<string, NodeDetails>
export interface NodeDetails {
  component: React.FC
  isDefault?: boolean
}

export interface NodeCollapsibleProps {
  /** parent element selector to listen resize event on */
  parentSelector: string
  /** percent child visible to collapse */
  percentageNodeVisible?: number
  /** margin from child bottom to start expanding */
  bottomMarginInPixels?: number
}
