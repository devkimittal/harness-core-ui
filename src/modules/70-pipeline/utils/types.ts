/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { StringNGVariable, NumberNGVariable, SecretNGVariable, PipelineInfoConfig } from 'services/cd-ng'
import type { NodeRunInfo, EntityGitDetails, EntityValidityDetails, InputSetResponse } from 'services/pipeline-ng'

export type AllNGVariables = StringNGVariable | NumberNGVariable | SecretNGVariable

export interface ExecutionPageQueryParams {
  view?: 'log' | 'graph'
  stage?: string
  step?: string
  retryStep?: string
}

export interface ExpressionBlock {
  expression?: string
  expressionValue?: string
}

export interface ConditionalExecutionNodeRunInfo extends NodeRunInfo {
  expressions?: ExpressionBlock[]
}

export type StatusType = 'SUCCESS' | 'FAILURE' | 'ERROR'

export type CreateUpdateInputSetsReturnType = Promise<{
  status?: StatusType
  nextCallback: () => void
}>

export interface InputSetType {
  name: string
  tags?: { [key: string]: string }
  identifier: string
  description: string
  orgIdentifier: string
  projectIdentifier: string
  pipeline: PipelineInfoConfig
  gitDetails: EntityGitDetails
  entityValidityDetails: EntityValidityDetails
}

export interface InputSetDTO extends Omit<InputSetResponse, 'identifier' | 'pipeline'> {
  pipeline?: PipelineInfoConfig
  identifier?: string
  repo?: string
  branch?: string
}

export interface SaveInputSetDTO {
  inputSet: InputSetDTO
}
