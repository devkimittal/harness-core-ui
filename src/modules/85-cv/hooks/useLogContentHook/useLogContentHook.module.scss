/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.dialog {
  padding: 0 !important;
  border-radius: 2px !important;
  width: fit-content;

  .healthSource {
    width: 225px;
  }
}

.fullScreen {
  width: 95vw !important;
  height: 70vh !important;
}

.main {
  min-width: 100%;
  width: 60vw;
  height: 60vh;
  overflow: hidden;
  color: var(--grey-300);
  background: var(--grey-900);

  pre.container {
    height: calc(100% - 70px);
  }

  .noLogs {
    padding: 0 15px;
    font-family: 'Roboto Mono', 'Monaco', 'Ubuntu Mono', monospace;
  }

  .text {
    font-family: 'Roboto Mono', 'Monaco', 'Ubuntu Mono', monospace !important;
  }

  .footer {
    position: absolute;
    left: 20px;
    right: 20px;
    bottom: 16px;

    .buttonLink {
      color: var(--primary-7);
      font-family: 'Roboto Mono', 'Monaco', 'Ubuntu Mono', monospace !important;
    }

    .button {
      padding: 8px !important;

      & > :global(.bp3-icon) {
        padding: 0 !important;
      }

      & > :global(.bp3-button-text) {
        width: 0;
        padding-left: 0;
        overflow: hidden;
        display: inline-block;
      }

      &:hover > :global(.bp3-button-text) {
        width: auto;
        padding-left: 4px;
      }
    }
  }
}
