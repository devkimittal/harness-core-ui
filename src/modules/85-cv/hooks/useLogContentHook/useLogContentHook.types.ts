/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SelectOption } from '@harness/uicore'
import type { State, UseActionCreatorReturn } from './views/ExecutionLog/ExecutionLog.types'

export interface UseLogContentHookProps {
  activityId: string
}

export interface UseLogContentHookReturn {
  openLogContentHook: () => void
  closeLogContentHook: () => void
}

export interface ExecutionLogHeaderProps {
  activityId: string
  healthSource: SelectOption
  setHealthSource: (healthSource: SelectOption) => void
  errorLogsOnly: boolean
  setErrorLogsOnly: (errorLogsOnly: boolean) => void
  actions: UseActionCreatorReturn
  setPageNumber: (pageNumber: number) => void
}

export interface ExecutionLogToolbarProps {
  state: State
  actions: UseActionCreatorReturn
  isFullScreen: boolean
  setIsFullScreen: (isFullScreen: React.SetStateAction<boolean>) => void
}

export enum LogTypes {
  ApiCallLog = 'ApiCallLog',
  ExecutionLog = 'ExecutionLog'
}
