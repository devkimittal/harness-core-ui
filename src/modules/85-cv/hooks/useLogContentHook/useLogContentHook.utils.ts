/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SelectOption } from '@harness/uicore'
import type { HealthSourceDTO } from 'services/cv'
import type { UseStringsReturn } from 'framework/strings'

export function isPositiveNumber(index: unknown): index is number {
  return typeof index === 'number' && index >= 0
}

export function getHealthSourceOptions(
  getString: UseStringsReturn['getString'],
  healthSources?: HealthSourceDTO[]
): SelectOption[] {
  const optionAll: SelectOption = { label: getString('all'), value: '' }

  const healthSourcesOptions =
    healthSources?.map(healthSource => ({
      label: healthSource.name ?? '',
      value: healthSource.identifier ?? ''
    })) ?? []

  return [optionAll, ...healthSourcesOptions]
}
