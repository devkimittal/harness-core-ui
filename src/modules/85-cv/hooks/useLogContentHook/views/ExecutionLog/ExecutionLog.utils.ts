/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { LogLineData } from '@pipeline/components/LogsContent/LogsState/types'
import type { LogLineData as ExecutionLog } from './ExecutionLog.types'

export function convertLogDataToLogLineData(data: ExecutionLog): LogLineData {
  const { text, searchIndices = {} } = data

  const _searchIndices = {
    level: searchIndices.logLevel,
    time: searchIndices.createdAt,
    out: searchIndices.log
  }

  switch (text.logLevel) {
    case 'WARN':
      return {
        text: {
          level: text.logLevel,
          time: text.createdAt,
          out: `\u001b[1;33m\u001b[40m${text.log}\u001b[0m`
        },
        searchIndices: _searchIndices
      }
    case 'ERROR':
      return {
        text: {
          level: `\u001b[1;31m\u001b[40m${text.logLevel}\u001b[0m`,
          time: text.createdAt,
          out: `\u001b[1;31m\u001b[40m${text.log}\u001b[0m`
        },
        searchIndices: _searchIndices
      }
    default:
      return {
        text: {
          level: text.logLevel,
          time: text.createdAt,
          out: text.log
        },
        searchIndices: _searchIndices
      }
  }
}
