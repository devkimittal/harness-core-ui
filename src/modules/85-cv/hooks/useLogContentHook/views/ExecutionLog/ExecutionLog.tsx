/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { Reducer } from 'react'
import cx from 'classnames'
import { useParams } from 'react-router-dom'
import { GroupedVirtuosoHandle, Virtuoso, VirtuosoHandle } from 'react-virtuoso'
import { Button, ButtonSize, ButtonVariation, Color, Container, Layout, SelectOption, Text } from '@harness/uicore'
import { useStrings, String as StrTemplate } from 'framework/strings'
import { useGetVerifyStepLogs } from 'services/cv'
import type { AccountPathProps } from '@common/interfaces/RouteInterfaces'
import { MultiLogLine } from '@pipeline/components/LogsContent/components/MultiLogLine/MultiLogLine'
import { getErrorMessage } from '@cv/utils/CommonUtils'
import ExecutionLogHeader from '../ExecutionLogHeader'
import ExecutionLogToolbar from '../ExecutionLogToolbar'
import { reducer, useActionCreator } from './ExecutionLogState'
import { defaultReducerState, PAGE_SIZE } from './ExecutionLog.constants'
import { isPositiveNumber } from '../../useLogContentHook.utils'
import { convertLogDataToLogLineData } from './ExecutionLog.utils'
import type { State, Action, ActionType, ExecutionLogProps } from './ExecutionLog.types'
import css from '../../useLogContentHook.module.scss'

const ExecutionLog: React.FC<ExecutionLogProps> = ({ activityId }) => {
  const { getString } = useStrings()
  const { accountId } = useParams<AccountPathProps>()
  const [state, dispatch] = React.useReducer<Reducer<State, Action<ActionType>>>(reducer, defaultReducerState)
  const actions = useActionCreator(dispatch)
  const [isFullScreen, setIsFullScreen] = React.useState(false)
  const [healthSource, setHealthSource] = React.useState<SelectOption>({ label: getString('all'), value: '' })
  const [errorLogsOnly, setErrorLogsOnly] = React.useState<boolean>(false)
  const [pageNumber, setPageNumber] = React.useState(0)
  const [isAtBottom, setIsAtBottom] = React.useState(false)
  const virtuosoRef = React.useRef<null | GroupedVirtuosoHandle | VirtuosoHandle>(null)
  const { currentIndex, linesWithResults } = state.searchData
  const length = state.data.length

  const { data, loading, error, refetch } = useGetVerifyStepLogs({
    verifyStepExecutionId: activityId,
    queryParams: {
      accountId,
      pageSize: PAGE_SIZE,
      logType: 'ExecutionLog',
      pageNumber,
      errorLogsOnly,
      ...(healthSource?.value ? { healthSources: [healthSource.value as string] } : {})
    },
    queryParamStringifyOptions: {
      arrayFormat: 'repeat'
    }
  })

  React.useEffect(() => {
    /* istanbul ignore else */ if (data?.resource?.content?.length) {
      actions.setExecutionLogs(data.resource.content)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data?.resource?.content])

  React.useEffect(() => {
    const index = linesWithResults[currentIndex]

    /* istanbul ignore next */
    if (virtuosoRef.current && isPositiveNumber(index)) {
      virtuosoRef.current.scrollToIndex(index)
    }
  }, [currentIndex, linesWithResults])

  const handleClick = (): void => {
    const handle = virtuosoRef.current
    /* istanbul ignore else */ if (handle) {
      handle.scrollToIndex(isAtBottom ? 0 : length)
    }
  }

  const errorMessage = getErrorMessage(error)
  const totalPages = data?.resource?.totalPages ?? 0

  return (
    <div>
      <ExecutionLogHeader
        activityId={activityId}
        healthSource={healthSource}
        setHealthSource={setHealthSource}
        errorLogsOnly={errorLogsOnly}
        setErrorLogsOnly={setErrorLogsOnly}
        setPageNumber={setPageNumber}
        actions={actions}
      />
      <ExecutionLogToolbar
        state={state}
        actions={actions}
        isFullScreen={isFullScreen}
        setIsFullScreen={setIsFullScreen}
      />
      <div className={cx(css.main, { [css.fullScreen]: isFullScreen })}>
        <pre className={css.container}>
          {length > 0 && (
            <Virtuoso
              height="100%"
              overscan={50}
              ref={virtuosoRef}
              atBottomThreshold={Math.ceil(length / 3)}
              atBottomStateChange={setIsAtBottom}
              totalCount={length}
              followOutput={_ => 'smooth'}
              itemContent={index => (
                <MultiLogLine
                  {...convertLogDataToLogLineData(state.data[index])}
                  limit={length}
                  lineNumber={index}
                  searchText={state.searchData.text}
                  currentSearchIndex={currentIndex}
                />
              )}
            />
          )}
          {!length && !loading && !errorMessage && (
            <StrTemplate tagName="div" className={css.noLogs} stringID="common.logs.noLogsText" />
          )}
        </pre>

        <div className={css.footer}>
          <Container flex>
            <Container width="80%">
              {loading && <StrTemplate tagName="div" className={css.text} stringID="loading" />}
              {!loading && errorMessage && (
                <Layout.Horizontal spacing="small">
                  <Text lineClamp={1} color={Color.RED_500} className={css.text}>
                    {errorMessage}
                  </Text>
                  <a className={css.buttonLink} onClick={() => refetch()}>
                    {getString('retry')}
                  </a>
                </Layout.Horizontal>
              )}
            </Container>
            {length > 0 && (
              <div>
                {pageNumber !== totalPages - 1 && (
                  <Button
                    text={getString('pipeline.verification.loadMore')}
                    icon="down"
                    className={css.button}
                    size={ButtonSize.SMALL}
                    margin={{ right: 'small' }}
                    variation={ButtonVariation.PRIMARY}
                    disabled={!!errorMessage}
                    onClick={() => setPageNumber(_prevPageNumber => _prevPageNumber + 1)}
                  />
                )}
                <Button
                  size={ButtonSize.SMALL}
                  variation={ButtonVariation.PRIMARY}
                  text={getString(isAtBottom ? 'cv.top' : 'cv.bottom')}
                  icon={isAtBottom ? 'arrow-up' : 'arrow-down'}
                  iconProps={{ size: 10 }}
                  onClick={handleClick}
                  className={css.button}
                />
              </div>
            )}
          </Container>
        </div>
      </div>
    </div>
  )
}

export default ExecutionLog
