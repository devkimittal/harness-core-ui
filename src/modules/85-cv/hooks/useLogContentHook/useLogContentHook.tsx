/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'
import { useModalHook } from '@harness/use-modal'
import { Classes } from '@blueprintjs/core'
import { Dialog } from '@harness/uicore'
import ExecutionLog from './views/ExecutionLog/ExecutionLog'
import type { UseLogContentHookProps, UseLogContentHookReturn } from './useLogContentHook.types'
import css from './useLogContentHook.module.scss'

export const useLogContentHook = ({ activityId }: UseLogContentHookProps): UseLogContentHookReturn => {
  const [showModal, hideModal] = useModalHook(
    () => (
      <Dialog isOpen enforceFocus={false} onClose={hideModal} className={cx(css.dialog, Classes.DIALOG)}>
        <ExecutionLog activityId={activityId} />
      </Dialog>
    ),
    [activityId]
  )

  return {
    openLogContentHook: showModal,
    closeLogContentHook: hideModal
  }
}
