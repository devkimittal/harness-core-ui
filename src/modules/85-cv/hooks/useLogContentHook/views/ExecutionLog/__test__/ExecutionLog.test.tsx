/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import userEvent from '@testing-library/user-event'
import { render, screen, waitFor } from '@testing-library/react'
import * as cvService from 'services/cv'
import { TestWrapper } from '@common/utils/testUtils'
import { LogTypes } from '@cv/hooks/useLogContentHook/useLogContentHook.types'
import ExecutionLog from '../ExecutionLog'
import {
  testWrapperProps,
  healthSourceResponse,
  executionLogsResponse,
  pathParams,
  errorMessage,
  deploymentActivitySummaryResponse
} from './ExecutionLog.mock'
import { PAGE_SIZE } from '../ExecutionLog.constants'

const STRING_ID_LOAD_MORE = 'pipeline.verification.loadMore'

jest.mock('services/cv', () => ({
  useGetVerifyStepHealthSources: jest
    .fn()
    .mockImplementation(() => ({ data: healthSourceResponse, loading: false, error: null, refetch: jest.fn() })),
  useGetVerifyStepLogs: jest
    .fn()
    .mockImplementation(() => ({ data: executionLogsResponse, loading: false, error: null, refetch: jest.fn() })),
  useGetVerifyStepDeploymentActivitySummary: jest.fn().mockImplementation(() => ({
    data: deploymentActivitySummaryResponse,
    loading: false,
    error: null,
    refetch: jest.fn()
  }))
}))

describe('ExecutionLog', () => {
  test('should render ExecutionLog component', () => {
    const { container } = render(
      <TestWrapper {...testWrapperProps}>
        <ExecutionLog activityId="activityId" />
      </TestWrapper>
    )

    expect(screen.getByText('cv.executionLogs')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('should call the api by clicking the load more button', async () => {
    render(
      <TestWrapper {...testWrapperProps}>
        <ExecutionLog activityId="activityId" />
      </TestWrapper>
    )

    expect(screen.getByRole('button', { name: STRING_ID_LOAD_MORE })).toBeInTheDocument()

    userEvent.click(screen.getByRole('button', { name: STRING_ID_LOAD_MORE }))

    await waitFor(() => {
      expect(cvService.useGetVerifyStepLogs).toHaveBeenLastCalledWith({
        verifyStepExecutionId: 'activityId',
        queryParams: {
          accountId: pathParams.accountId,
          pageSize: PAGE_SIZE,
          logType: LogTypes.ExecutionLog,
          pageNumber: 1,
          errorLogsOnly: false
        },
        queryParamStringifyOptions: {
          arrayFormat: 'repeat'
        }
      })
      expect(screen.queryByRole('button', { name: STRING_ID_LOAD_MORE })).not.toBeInTheDocument()
    })
  })

  test('should handle the Top and Bottom navigation', () => {
    const { container } = render(
      <TestWrapper {...testWrapperProps}>
        <ExecutionLog activityId="activityId" />
      </TestWrapper>
    )

    expect(container.querySelector('[data-icon="arrow-down"]')).toBeInTheDocument()

    userEvent.click(container.querySelector('[data-icon="arrow-down"]')!)
  })

  test('should show the text Loading...', () => {
    jest
      .spyOn(cvService, 'useGetVerifyStepLogs')
      .mockReturnValue({ data: {}, loading: true, error: null, refetch: jest.fn() } as any)

    render(
      <TestWrapper {...testWrapperProps}>
        <ExecutionLog activityId="activityId" />
      </TestWrapper>
    )

    expect(screen.getByText('loading')).toBeInTheDocument()
  })

  test('should show the error message and fetch the logs by clicking on Retry', async () => {
    jest
      .spyOn(cvService, 'useGetVerifyStepLogs')
      .mockReturnValue({ data: {}, loading: false, error: { message: errorMessage }, refetch: jest.fn() } as any)

    render(
      <TestWrapper {...testWrapperProps}>
        <ExecutionLog activityId="activityId" />
      </TestWrapper>
    )

    expect(screen.getByText(errorMessage)).toBeInTheDocument()

    userEvent.click(screen.getByText('retry'))

    await waitFor(() => {
      expect(cvService.useGetVerifyStepLogs).toHaveBeenLastCalledWith({
        verifyStepExecutionId: 'activityId',
        queryParams: {
          accountId: pathParams.accountId,
          pageSize: PAGE_SIZE,
          logType: LogTypes.ExecutionLog,
          pageNumber: 0,
          errorLogsOnly: false
        },
        queryParamStringifyOptions: {
          arrayFormat: 'repeat'
        }
      })
    })
  })
})
