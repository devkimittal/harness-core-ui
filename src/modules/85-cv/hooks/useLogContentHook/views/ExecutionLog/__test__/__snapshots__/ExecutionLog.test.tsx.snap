// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutionLog should render ExecutionLog component 1`] = `
<div>
  <div>
    <div>
      <h2
        class="StyledProps--font StyledProps--font-h2 StyledProps--main StyledProps--font-variation-form-title StyledProps--padding-xlarge StyledProps--border StyledProps--border-bottom"
        level="2"
      >
        cv.executionLogs
      </h2>
      <div
        class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main StyledProps--padding-xlarge"
      >
        <div
          class="StyledProps--font StyledProps--main"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-tiny-semi StyledProps--color StyledProps--color-grey400"
          >
            connectors.cdng.monitoredService.label
          </p>
          <span
            class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-black StyledProps--font-variation-tiny StyledProps--font-weight-bold StyledProps--padding-right-xsmall"
          >
            dynatrace
          </span>
          <span
            class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--font-variation-tiny-semi StyledProps--color StyledProps--color-grey700"
          >
            prod
          </span>
        </div>
        <div
          class="StyledProps--font StyledProps--main StyledProps--border StyledProps--border-left"
        />
        <div
          class="bp3-popover-wrapper healthSource Select--main bp3-fill"
        >
          <div
            class="bp3-popover-target"
          >
            <div
              class="bp3-input-group"
            >
              <input
                autocomplete="off"
                class="bp3-input"
                placeholder="- Select -"
                style="padding-right: 0px;"
                type="text"
                value="pipeline.verification.healthSourceLabel: all"
              />
              <span
                class="bp3-input-action"
              >
                <span
                  class="bp3-icon bp3-icon-chevron-down StyledProps--main StyledProps--padding-small"
                  icon="chevron-down"
                >
                  <svg
                    data-icon="chevron-down"
                    height="14"
                    viewBox="0 0 16 16"
                    width="14"
                  >
                    <desc>
                      chevron-down
                    </desc>
                    <path
                      d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
        <label
          class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox StyledProps--main"
        >
          <input
            type="checkbox"
          />
          <span
            class="bp3-control-indicator"
          />
          cv.displayOnlyErrors
        </label>
      </div>
    </div>
    <div
      class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--padding-small StyledProps--background-primary10 StyledProps--border StyledProps--border-bottom StyledProps--border-color-grey700"
    >
      <div />
      <div
        class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
      >
        <div
          class="bp3-input-group ui-search-box ExpandingSearchInput--main  ExpandingSearchInput--flip  ExpandingSearchInput--dark"
          data-name=""
        >
          <span
            class="bp3-icon StyledProps--main ExpandingSearchInput--icon StyledProps--color StyledProps--color-grey200"
            data-icon="thinner-search"
          >
            <svg
              fill="none"
              height="14"
              viewBox="0 0 16 16"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.125 13.25c1.617 0 3.32-.695 4.375-1.75L15 15l-3.5-3.5c1.021-1.005 1.75-2.894 1.75-4.375a6.125 6.125 0 10-6.125 6.125z"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
          <input
            class="bp3-input  ExpandingSearchInput--find"
            dir="auto"
            placeholder="Search"
            style="padding-right: 0px;"
            type="search"
            value=""
          />
        </div>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
          type="button"
        >
          <span
            class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey200"
            data-icon="full-screen"
          >
            <svg
              fill="none"
              height="22"
              viewBox="0 0 24 24"
              width="22"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.12 19.512h4.876v-4.815M8.88 19.512H4.003v-4.815M15.12 4.488h4.876v4.814M8.88 4.488H4.003v4.814"
                stroke="currentColor"
                stroke-linecap="round"
              />
            </svg>
          </span>
        </button>
        <div
          class="StyledProps--font StyledProps--main StyledProps--border StyledProps--border-left StyledProps--border-color-grey700"
        />
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-link"
          type="button"
        >
          <span
            class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
            data-icon="command-install"
          >
            <svg
              height="16"
              viewBox="0 0 14 14"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.4 10.2c.2.2.4.3.6.3s.5-.1.6-.3l3.5-3.5c.2-.2.3-.4.3-.6 0-.5-.4-.9-.9-.9-.2 0-.5.1-.6.3l-2 2V.9C7.9.4 7.5 0 7 0s-.9.4-.9.9v6.6l-2-2c-.1-.2-.4-.3-.6-.3-.5 0-.9.4-.9.9 0 .2.1.5.3.6zm6.7-.6c-.5 0-.9.4-.9.9v1.8H1.8v-1.8c0-.5-.4-.9-.9-.9s-.9.4-.9.9v2.6c0 .5.4.9.9.9h12.2c.5 0 .9-.4.9-.9v-2.6c0-.5-.4-.9-.9-.9z"
                fill="#4d5969"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            cv.download
          </span>
        </button>
      </div>
    </div>
    <div
      class="main"
    >
      <pre
        class="container"
      >
        <div
          data-test-id="virtuoso-scroller"
          data-virtuoso-scroller="true"
          height="100%"
          style="height: 100%; outline: none; overflow-y: auto; position: relative;"
          tabindex="0"
        >
          <div
            data-viewport-type="element"
            style="width: 100%; height: 100%; position: absolute; top: 0px;"
          >
            <div
              data-test-id="virtuoso-item-list"
              style="box-sizing: border-box; padding-top: 0px; padding-bottom: 0px; margin-top: 0px;"
            />
          </div>
        </div>
      </pre>
      <div
        class="footer"
      >
        <div
          class="StyledProps--font StyledProps--main StyledProps--flex"
        >
          <div
            class="StyledProps--font StyledProps--main"
            style="width: 80%;"
          />
          <div>
            <button
              class="bp3-button Button--button StyledProps--font StyledProps--main button StyledProps--margin-right-small Button--with-current-color Button--withLeftIcon Button--variation Button--variation-primary Button--size Button--small"
              type="button"
            >
              <span
                class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
                data-icon="down"
              >
                <svg
                  height="16"
                  viewBox="0 0 14 10"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1.207 6.125l5.78 2.75 5.806-2.75m-11.586-5l5.78 2.75 5.806-2.75"
                    fill="none"
                    stroke="#3E7BFA"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
              <span
                class="bp3-button-text"
              >
                pipeline.verification.loadMore
              </span>
            </button>
            <button
              class="bp3-button Button--button StyledProps--font StyledProps--main button Button--with-current-color Button--withLeftIcon Button--variation Button--variation-primary Button--size Button--small"
              type="button"
            >
              <span
                class="bp3-icon bp3-icon-arrow-down StyledProps--main StyledProps--padding-right-xsmall"
                icon="arrow-down"
              >
                <svg
                  data-icon="arrow-down"
                  height="10"
                  viewBox="0 0 16 16"
                  width="10"
                >
                  <desc>
                    arrow-down
                  </desc>
                  <path
                    d="M13 8c-.3 0-.5.1-.7.3L9 11.6V2c0-.5-.4-1-1-1s-1 .5-1 1v9.6L3.7 8.3C3.5 8.1 3.3 8 3 8c-.5 0-1 .5-1 1 0 .3.1.5.3.7l5 5c.2.2.4.3.7.3s.5-.1.7-.3l5-5c.2-.2.3-.4.3-.7 0-.6-.4-1-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="bp3-button-text"
              >
                cv.bottom
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
